### ISSUE TITLE
Add CMake option to build Valhalla with `-fno-rtti`

### ISSUE THREAD

**Issue by SiarheiFedartsou** [valhalla/valhalla#3228]
Some projects depending on Valhalla have a requirement to be built with `-fno-rtti` flag(for various reasons, mainly due to binary size). In general there are no any issues with Valhalla and `-fno-rtti`, but from time to time new changes are introduced in Valhalla which explicitly require RTTI: e.g. those ones which use `typeid` or `dynamic_cast`(https://github.com/valhalla/valhalla/pull/3191). 
The idea is the following:
1. Introduce CMake option `DISABLE_RTTI`, which applies `-fno-rtti` if `ON`.
2. Configure CI job which would check if Valhalla can be built with `DISABLE_RTTI`.

cc @valhalla/core

**Comment by kevinkreiser** [comment]
@SiarheiFedartsou sorry it didnt occur to me that you guys were disabling rtti. it seems to me that having a cmake flag control it doesnt make sense because even if CI can detect it you'll just have to rewrite your code to work around it. if its a hard requirement downstream we should just avoid it completely and make impossible to opt out of it. before we do that i would ask you, how much does it actually save on binary size?

**Comment by SiarheiFedartsou** [comment]
@kevinkreiser 
also I see we actively use `boost::property_tree` which requires RTTI. [There is PR in Boost which fixes this](https://github.com/boostorg/property_tree/pull/20), but it is being stuck for 5 years. Don't we want to get rid of this and use something like usual json or strong-typed structure instead? WDYT? I could take this work probably.

**Comment by SiarheiFedartsou** [comment]
> @kevinkreiser also I see we actively use `boost::property_tree` which requires RTTI. [There is PR in Boost which fixes this](https://github.com/boostorg/property_tree/pull/20), but it is being stuck for 5 years. Don't we want to get rid of this and use something like usual json or strong-typed structure instead? WDYT? I could take this work probably.

if you wonder how do we build Valhalla without RTTI now: we simply use patched version of boost, but want to switch to mainstream

**Comment by SiarheiFedartsou** [comment]
> also protobuf is loaded with typeid, at least the version im compiling that comes with my system. iirc you build protobuf from source and disable rtti for it directly?

Yes, we use `GOOGLE_PROTOBUF_NO_RTTI`

**Comment by kevinkreiser** [comment]
Yes we would like to stop using ptree and use rapidjson everywhere but it's a decent amount of work which is why we haven't already done it. If you are interested in doing some of it that would be great
